pms_website_sale/controllers/booking_engine_parser.py

```diff
@@ -0,0 +1,165 @@
+                return room_request
+        return None
+
+    def parse(self):
+        """Create a booking.engine based on the parser data"""
+        values = self._get_booking_engine_vals()
+        self.booking_engine = self.env["pms.booking.engine"].sudo().create(values)
+        self._populate_availability_results()
+        return self.booking_engine
+
+    def save(self):
+        """Save data into session"""
+        self._session[BookingEngineParser.SESSION_KEY] = self.data
+
+    def set_daterange(self, start_date, end_date, overwrite=True):
+        """Set a start_date and a end_date for booking"""
+        if (
+            not self.data.get("start_date")
+            and not self.data.get("end_date")
+            or overwrite
+        ):
+            BookingEngineParser._check_daterange(start_date, end_date)
+            self.data["start_date"] = str(start_date) if start_date else None
+            self.data["end_date"] = str(end_date) if end_date else None
+
+    def add_room_request(self, room_type_id, quantity, start_date=None, end_date=None):
+        """Add or update a room request to the booking"""
+        # FIXME: Perhaps this method should be renamed
+        # `update_room_request` because it also update records when a
+        # record already exists. Or it can be split into two function.
+        # An idea can be also to merge `del_room_request` and
+        # `add_room_request` into `update_room_request` where the
+        # deletion is performed when quantity is null.
+        # To be discussed.
+        if start_date != self.data.get("start_date") or end_date != self.data.get(
+            "end_date"
+        ):
+            ParserError(
+                "Booking date does not match existing booking date. "
+                "The room cannot be booked."
+            )
+        try:
+            room_type_id = int(room_type_id)
+            quantity = int(quantity)
+        except (ValueError, TypeError):
+            raise ParserError("Cannot proceed due to incorrect value.")
+
+        new_room_request = {
+            "room_type_id": room_type_id,
+            "quantity": quantity,
+        }
+        existing_room_request = self._get_room_request(room_type_id)
+        if existing_room_request:
+            existing_room_request.update(new_room_request)
+        else:
+            self.data["rooms_requests"].append(new_room_request)
+
+    def del_room_request(self, room_type_id):
+        """Remove a room request to the booking"""
+        try:
+            room_type_id = int(room_type_id)
+        except (ValueError, TypeError):
+            raise ParserError("Cannot proceed due to incorrect value")
+        del_index = None
+        for index, room_request in enumerate(self.data["rooms_requests"]):
+            if room_request.get("room_type_id") == room_type_id:
+                del_index = index
+        if del_index is not None:
+            self.data["rooms_requests"].pop(del_index)
```

Comment:
why not rather ?

```suggestion
        rooms_requests = self.data["rooms_requests"]
        rooms_requests = [rr for rr in rooms_requests if rr[room_type_id] != room_type_id]
        self.data["rooms_requests"] = rooms_requests
```
or
```suggestion
        self.data["rooms_requests"] = list(filter(
            lambda rr: rr[room_type_id] != room_type_id, self.data["rooms_requests"]
        ))
```

Reply:
Indeed nice catch. :)